File: src/main/java/com/extendedae_plus/mixin/impl/HolderCardAutoCompletionState.java
package com.extendedae_plus.mixin.impl;

import appeng.api.crafting.IPatternDetails;
import appeng.api.networking.IManagedGridNode;
import appeng.api.networking.crafting.ICraftingService;
import appeng.api.upgrades.IUpgradeInventory;
import appeng.me.cluster.implementations.CraftingCPUCluster;
import com.extendedae_plus.common.init.ModItems;
import com.extendedae_plus.mixin.bridge.HelperCraftingJob;
import com.extendedae_plus.mixin.core.advancedae.accessor.AccessorAdvCraftingCPULogic;
import com.extendedae_plus.mixin.core.advancedae.accessor.AccessorAdvExecutingCraftingJob;
import com.extendedae_plus.mixin.core.ae2.accessor.AccessorCraftingCPULogic;
import com.extendedae_plus.mixin.core.ae2.accessor.AccessorExecutingCraftingJob;
import net.minecraft.util.Tuple;
import net.pedroksl.advanced_ae.common.cluster.AdvCraftingCPU;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Supplier;

public class HolderCardAutoCompletionState {
    private final IManagedGridNode mainNode;
    private final Supplier<IUpgradeInventory> getterUpgradeInventory;

    private boolean cardAvailable = false;

    public static final HolderCardAutoCompletionState EMPTY = new HolderCardAutoCompletionState(
            null, null) {
        @Override public boolean cardAvailable() {
            return false;
        }
        @Override public void onUpgradesChanged() {}
    };

    public HolderCardAutoCompletionState(IManagedGridNode mainNode, Supplier<IUpgradeInventory> getterUpgradeInventory) {
        this.mainNode = mainNode;
        this.getterUpgradeInventory = getterUpgradeInventory;
    }

    public boolean cardAvailable() {
        return this.cardAvailable;
    }

    public void onUpgradesChanged() {
        var upgradesInv = this.getterUpgradeInventory.get();
        if (upgradesInv == null) return;
        var state = new AtomicBoolean(false);
        upgradesInv.forEach(card -> {
            if (!state.get() && card.is(ModItems.CARD_AUTO_COMPLETION))
                state.set(true);
        });
        this.cardAvailable = state.get();
    }

    public void completeJob(IPatternDetails patternDetails) {
        if (!this.cardAvailable()) return;
        if (patternDetails == null) return;

        var node = this.mainNode.getNode();
        if (node == null) return;
        var grid = node.getGrid();
        if (grid == null) return;
        var craftingService = grid.getCraftingService();
        if (craftingService == null) return;

        var taskResult = getTasks(craftingService);
        var tasks = taskResult.getB();
        if (tasks.isEmpty()) return;

        var progress = tasks.get(patternDetails);
        if (progress == null) {
            for (var progressEntry : tasks.entrySet()) {
                if (!testProgress(patternDetails, progressEntry.getKey())) continue;
                progress = progressEntry.getValue();
                break;
            }
        }

        if (progress == null) return;
        if (progress.getValue() > 1) return;
        taskResult.getA().run();
    }

    private static Tuple<Runnable, Map<IPatternDetails, HelperCraftingJob.HelperJobProgress>>
    getTasks(ICraftingService craftingService) {
        Tuple<Runnable, Map<IPatternDetails, HelperCraftingJob.HelperJobProgress>> result =
                new Tuple<>(() -> {}, new HashMap<>());
        craftingService.getCpus().forEach(craftingCPU -> {
            if (!craftingCPU.isBusy()) return;

            result.setA(craftingCPU::cancelJob);
            if (craftingCPU instanceof CraftingCPUCluster cpu
                    && cpu.craftingLogic instanceof AccessorCraftingCPULogic accessorLogic
                    && accessorLogic.getJob() instanceof AccessorExecutingCraftingJob accessorJob) {
                result.getB().putAll(accessorJob.getTasks());
            } else if (craftingCPU instanceof AdvCraftingCPU cpu
                    && cpu.craftingLogic instanceof AccessorAdvCraftingCPULogic accessorLogic
                    && accessorLogic.getJob() instanceof AccessorAdvExecutingCraftingJob accessorJob) {
                result.getB().putAll(accessorJob.getTasks());
            }
        });
        return result;
    }

    private static boolean testProgress(IPatternDetails targetPattern, IPatternDetails taskPattern) {
        if (taskPattern == targetPattern) return true;
        if (taskPattern != null && targetPattern != null) {
            var taskDefinition = taskPattern.getDefinition();
            return taskDefinition != null && taskDefinition.equals(targetPattern.getDefinition());
        }
        return false;
    }
}

File: src/main/kotlin/com/extendedae_plus/mixin/impl/HolderCardAutoCompletionState.kt
package com.extendedae_plus.mixin.impl

import appeng.api.crafting.IPatternDetails
import appeng.api.features.IPlayerRegistry
import appeng.api.networking.IManagedGridNode
import appeng.api.networking.crafting.ICraftingCPU
import appeng.api.upgrades.IUpgradeInventory
import appeng.core.network.clientbound.CraftingJobStatusPacket
import appeng.me.cluster.implementations.CraftingCPUCluster
import com.extendedae_plus.common.init.EAEPItems
import com.extendedae_plus.mixin.core.advancedae.accessor.AccessorCraftingLogicAdv
import com.extendedae_plus.mixin.core.ae2.accessor.AccessorCraftingLogic
import com.extendedae_plus.mixin.helper.HelperCraftingJob
import com.fish.fishlib.network.base.PacketGeneric.Companion.sendToPlayer
import net.minecraft.util.Tuple
import net.minecraft.world.level.block.entity.BlockEntity
import net.pedroksl.advanced_ae.common.cluster.AdvCraftingCPU
import java.util.concurrent.atomic.AtomicBoolean

open class HolderCardAutoCompletionState(
    private val mainNode: IManagedGridNode?,
    private val tile: () -> BlockEntity?,
    private val inventoryUpgrade: () -> IUpgradeInventory?
) {
    var cardAvailable = false
        private set

    open fun onUpgradesChanged() {
        val upgradesInv = this.inventoryUpgrade() ?: return
        val state = AtomicBoolean(false)
        upgradesInv.forEach { card ->
            if (!state.get() && card.`is`(EAEPItems.CardAutoCompletion))
                state.set(true)
        }
        this.cardAvailable = state.get()
    }

    fun completeJob(pattern: IPatternDetails) = this.taskProcess { canceler, helper ->
        val infoTask = helper.tasks

        val progress = infoTask.entries
            .find { pattern == it.key }
            ?.value
            ?.value
            ?: return@taskProcess
        if (progress > 1) return@taskProcess

        canceler()
    }

    private fun taskProcess(
        processor: (() -> Unit, HelperCraftingJob) -> Unit
    ) {
        if (!this.cardAvailable) return

        val serviceCrafting = this.mainNode?.grid?.craftingService ?: return

        val contextProcess = Tuple<(() -> Unit)?, HelperCraftingJob?>(null, null)
        serviceCrafting.cpus
            .filter(ICraftingCPU::isBusy)
            .forEach { cpu ->
                contextProcess.a = cpu::cancelJob
                contextProcess.b =
                    ((cpu as? CraftingCPUCluster)?.craftingLogic as? AccessorCraftingLogic)?.job as? HelperCraftingJob
                        ?: ((cpu as? AdvCraftingCPU)?.craftingLogic as? AccessorCraftingLogicAdv)?.job as? HelperCraftingJob
                                ?: return@forEach
            }
        val canceler = contextProcess.a ?: return
        val helper = contextProcess.b ?: return
        processor(canceler, helper)

        val player = IPlayerRegistry.getConnected(
            (tile()?.level?.server ?: return), helper.playerID ?: return
        ) ?: return
        CraftingJobStatusPacket(
            helper.link.craftingID,
            helper.outputFinal.what,
            helper.outputFinal.amount,
            helper.remainingAmount,
            CraftingJobStatusPacket.Status.FINISHED
        ).sendToPlayer(player)
    }

    companion object {
        @JvmField
        val Empty: HolderCardAutoCompletionState = object : HolderCardAutoCompletionState(
            null, { null }, { null }
        ) {
            override fun onUpgradesChanged() = Unit
        }
    }
}
